<!-- MUSIAL (Webserver) | Simon Hackl - simon.hackl@uni-tuebingen.de | v1.2.0 | GPL-3.0 license | github.com/Integrative-Transcriptomics/MUSIAL-WEB -->

{% extends 'index.html' %} {% block styles %}
<link
  rel="stylesheet"
  type="text/css"
  href="{{url_for('static', filename='css/upload.css')}}"
/>
{% endblock %} {% block scripts %}
<script
  type="text/javascript"
  src="{{url_for('static', filename='js/papaparse.min.js')}}"
  defer
></script>
<script
  type="text/javascript"
  src="{{url_for('static', filename='js/upload.js')}}"
  defer
></script>
{% endblock %} {% block menu %}
<!-- Menu elements floating left -->
<div class="app-bar-container mr-auto d-none d-flex-fs">
  <a
    id="menu-link-home"
    href="{{url_for('get_template_home')}}"
    class="app-bar-item text-upper"
    >home</a
  >
  <a
    id="menu-link-upload"
    href="{{url_for('get_template_upload')}}"
    class="app-bar-item text-upper active"
    >upload</a
  >
  <a
    id="menu-link-results"
    href="{{url_for('get_template_results')}}"
    class="app-bar-item text-upper"
    >results</a
  >
  <a
    id="menu-link-upload"
    href="{{url_for('get_template_help')}}"
    class="app-bar-item text-upper"
    >help</a
  >
  <a
    id="menu-link-legal"
    href="{{url_for('get_template_legal')}}"
    class="app-bar-item text-upper"
    >legal</a
  >
</div>
<!-- Menu elements floating right -->
<div class="app-bar-container ml-auto d-none d-flex-fs">
  <i id="menu-active-session-indicator" class="p-2 fa-duotone fa-hexagon"></i>
</div>
{% endblock %} {% block content %}
<div id="main-upload" class="main">
  <div class="remark m-4">
    <h3 class="p-2 text-left">
      <small><i class="fa-solid fa-dna"></i> Reference Sequence </small>
      <i
        id="collapse-reference-genome"
        class="fa-solid fa-circle-info info-clickable float-right"
      ></i>
    </h3>
    <div
      class="info-block mb-6"
      data-role="collapse"
      data-toggle-element="#collapse-reference-genome"
      data-collapsed="true"
    >
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent">reference sequence</span>
          <a
            href="https://www.ncbi.nlm.nih.gov/genbank/fastaformat/"
            target="_blank"
            >FASTA</a
          >
          format file containing one or more contigs/chromosomes/plasmids to use
          as the reference sequence for the analysis. Ensure that the sequence
          identifier (SeqID) is unique for each entry and matches the
          identifiers in the used variant call (.vcf) and generic feature format
          (.gff) files. <b>Format example:</b>
        </p>
        <p class="tag-info-text">
          >NC_000962.3 Mycobacterium tuberculosis H37Rv, complete genome
          <br />TTGACCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACGCGGTCGTCTCCGAACTTAACGGCGACC
          <br />CTAAGGTTGACGACGGACCCAGCAGTGATGCTAATCTCAGCGCTCCGCTGACCCCTCAGCAAAGGGCTTG
          <br />
        </p>
      </div>
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent">excluded positions (optional)</span>
          Specifies positions to be excluded from the analysis. Entries can be
          added as tags in the input field by pressing enter.
          <b>Format example:</b>
        </p>
        <p class="tag-info-text"><span class="tag-citation">NC_000962.3 1473246,12468-13016</span></p>
      </div>
    </div>
    <input
      id="input-reference-genome"
      type="file"
      data-role="file"
      data-button-title='<i class="fa-solid fa-file-arrow-up"></i>'
      data-prepend="<small><span class='tag-light'>reference sequence</span> (.fasta)<sup class='mandatory-tag'>*</sup></small>"
      accept=".fasta,.fasta"
    />
    <br>
    <label for="input-reference-genome-excluded-positions">
      <small><span class='tag-light mt-4'>excluded positions</span></small>
    </label>
    <input
      id="input-reference-genome-excluded-positions"
      class="mt-2"
      type="text"
      data-role="taginput"
      data-tag-trigger="Enter"
    >
    </input>
  </div>

  <div class="remark m-4">
    <h3 class="p-2 text-left">
      <small><i class="fa-solid fa-bacteria"></i> Samples</small>
      <i
        id="collapse-samples"
        class="fa-solid fa-circle-info info-clickable float-right"
      ></i>
    </h3>
    <div
      class="info-block mb-6"
      data-role="collapse"
      data-toggle-element="#collapse-samples"
      data-collapsed="true"
    >
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent">variant calls</span> One or more
          files in
          <a
            href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531692-VCF-Variant-Call-Format"
            target="_blank"
            >VCF</a
          >
          format, each containing variant calls for exactly one biological
          sample. If haploid calls are provided, only the variant with the
          highest frequency in terms of number of reads is used. Ensure that the
          CHROM column's row values match any SeqID of the provided reference
          genome. Each filename will be used as the internal sample identifier.
          Required fields can be derived from the example below. <b>Format example:</b>
        </p>
        <p class="tag-info-text">
          #CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	ERR017782_pass
          <br />
          NC_000962.3	1534	.	T	C	1148.06	.	AC=2;AF=1.00;AN=2;DP=38;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=34.79;SOR=1.542	GT:AD:DP:GQ:PL	1/1:0,33:33:99:1162,99,0
          <br />
          NC_000962.3	1977	.	A	G	781.06	.	AC=2;AF=1.00;AN=2;DP=24;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.88;QD=25.36;SOR=2.200	GT:AD:DP:GQ:PL	1/1:0,21:21:63:795,63,0
        </p>
      </div>
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent">sample meta information (optional)</span>
          A file in
          <a href="https://docs.fileformat.com/spreadsheet/tsv/" target="_blank"
            >TSV</a
          >
          format format containing meta information to be added to sample
          entries. Column names and cell values can be arbitrary; the first
          column must contain sample identifiers that match a provided variant
          call file. <b>Format example:</b>
        </p>
        <p class="tag-info-text">
          accession	strain	date	country	region	lineage
          <br />
          ERR970409	G04126	1970-01-02	South Africa	Africa	D
          <br />
          ERR234106	G00883	1970-01-02	Germany	Europe	L6
        </p>
      </div>
    </div>
    <input
      id="input-samples-files"
      type="file"
      data-role="file"
      data-button-title='<i class="fa-solid fa-file-arrow-up"></i>'
      data-prepend="<small><span class='tag-light'>variant calls</span> (.vcf)<sup class='mandatory-tag'>*</sup></small>"
      accept=".vcf"
      multiple
    />
    <input
      id="input-samples-meta"
      type="file"
      data-role="file"
      data-button-title='<i class="fa-solid fa-file-arrow-up"></i>'
      data-prepend="<small><span class='tag-light'>sample meta information</span> (.tsv)</small>"
      class="mt-4"
      accept=".csv,.tsv"
    />
  </div>

  <div class="remark m-4">
    <h3 class="p-2 text-left">
      <small
        ><i class="fa-solid fa-magnifying-glass"></i> Genomic Features</small
      >
      <i
        id="collapse-features"
        class="fa-solid fa-circle-info info-clickable float-right"
      ></i>
    </h3>
    <div
      class="info-block mb-6"
      data-role="collapse"
      data-toggle-element="#collapse-features"
      data-collapsed="true"
    >
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent">reference sequence annotation</span> A file in
          <a
            href="https://www.ensembl.org/info/website/upload/gff.html?redirect=no"
            target="_blank"
            >GFF</a
          >
          format, describing features of the provided reference sequence. Ensure
          that the seqname cell entries match the SeqID values of the provided
          reference genome. <b>Format example:</b>
        </p>
        <p class="tag-info-text">
          NC_000962.3	RefSeq	region	1	4411532	.	+	.	ID=NC_000962.3:1..4411532;Dbxref=taxon:83332;gbkey=Src;genome=genomic;mol_type=genomic DNA;strain=H37Rv;type-material=type strain of Mycobacterium tuberculosis
          <br>
          NC_000962.3	RefSeq	gene	1	1524	.	+	.	ID=gene-Rv0001;Dbxref=GeneID:885041;Name=dnaA;experiment=DESCRIPTION:Mutation analysis%2C gene expression[PMID: 10375628];gbkey=Gene;gene=dnaA;gene_biotype=protein_coding;locus_tag=Rv0001
        </p>
      </div>
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent">features to analyze</span>
          One or more features to be analyzed. The 
          feature information is extracted from the provided
          reference sequence annotation file; To assign a feature, a
          corresponding attribute key and value of the annotation
          file must be provided. Entries can be added as tags in
          the input field by pressing enter. Full genome analysis
          can be run by specifying whole regions as features.
          <b>Format example:</b>
        </p>
        <p class="tag-info-text"><span class="tag-citation">Name katG</span>&nbsp;<span class="tag-citation">gbkey Src</span></p>
      </div>
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent">feature meta information (optional)</span>
          A file in
          <a href="https://docs.fileformat.com/spreadsheet/tsv/" target="_blank"
            >TSV</a
          >
          format containing meta information to be added to gene/feature
          entries. Column names and cell values can be arbitrary; the first
          column must contain gene identifiers that match a specified gene to
          analyze. <b>Format example:</b>
        </p>
        <p class="tag-info-text">
          name product molecular_function amr_relation
          <br />
          katG Catalase-peroxidase antioxidant activity,oxidoreductase activity isoniazid
        </p>
      </div>
    </div>
    <input
      id="input-features-file"
      type="file"
      data-role="file"
      data-button-title='<i class="fa-solid fa-file-arrow-up"></i>'
      data-prepend="<small><span class='tag-light'>reference sequence annotation</span> (.gff)<sup class='mandatory-tag'>*</sup></small>"
      accept=".gff,.gff3,.gff,gff3"
    />
    <br>
    <label for="input-features-list">
      <small><span class='tag-light'>features to analyze</span><sup class='mandatory-tag'>*</sup></small>
    </label>
    <input
      id="input-features-list"
      class="mt-2"
      type="text"
      data-role="taginput"
      data-on-tag-input-create="observeFeaturesList"
      data-tag-trigger="Enter"
    ></input>
    <input
      id="input-features-meta"
      type="file"
      data-role="file"
      data-button-title='<i class="fa-solid fa-file-arrow-up"></i>'
      data-prepend="<small><span class='tag-light'>feature meta information</span> (.tsv)</small>"
      class="mt-4"
      accept=".csv,.tsv"
    />
  </div>

  <div class="remark m-4">
    <h3 class="p-2 text-left">
      <small
        ><i class="fa-solid fa-magnifying-glass"></i> Proteoform Analysis</small
      >
      <i
        id="collapse-protein"
        class="fa-solid fa-circle-info info-clickable float-right"
      ></i>
    </h3>
    <div
      class="info-block mb-6"
      data-role="collapse"
      data-toggle-element="#collapse-protein"
      data-collapsed="true"
    >
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent">proteoform analysis</span>
          Set this option to infer aminoacid variants of specified features
          independent of provided
          <a
            href="https://www.wwpdb.org/documentation/file-format-content/format33/v3.3.html"
            target="_blank"
            >PDB</a
          >
          files. <b>This option should only be set, if all specified features are coding genes.</b>
        </p>
      </div>
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent">protein structures (optional)</span>
          One or multiple files in
          <a
            href="https://www.wwpdb.org/documentation/file-format-content/format33/v3.3.html"
            target="_blank"
            >PDB</a
          >
          format. The file names have to match with specified features (values) to analyze
          and the corresponding structure will be assigned to the respective
          feature. Providing protein structures will trigger the inference of
          aminoacid variants and proteoforms of the respective feature. <b>Format example:</b>
        </p>
        <p class="tag-info-text">
          HEADER
          <br>
          ATOM&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;N&nbsp;&nbsp;&nbsp;MET&nbsp;A&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;42.903&nbsp;&nbsp;&nbsp;4.355&nbsp;&nbsp;&nbsp;0.110&nbsp;&nbsp;1.00&nbsp;28.05&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;N  
          <br>
          ATOM&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2&nbsp;&nbsp;CA&nbsp;&nbsp;MET&nbsp;A&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;42.216&nbsp;&nbsp;&nbsp;5.583&nbsp;&nbsp;-0.321&nbsp;&nbsp;1.00&nbsp;28.05&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;C  
        </p>
      </div>
    </div>
    <input
      id="input-infer-proteoforms"
      type="checkbox"
      class="mt-4"
      checked
      data-role="switch"
      data-caption="Infer Proteoform Information"
      data-cls-switch="custom-switch-on-off"
      data-material="true"
    />
    <input
      id="input-features-proteins"
      type="file"
      data-role="file"
      data-button-title='<i class="fa-solid fa-file-arrow-up"></i>'
      data-prepend="<small><span class='tag-light'>protein structures</span> (.pdb)</small>"
      class="mt-4"
      accept=".pdb"
      multiple
    />
  </div>

  <div class="remark m-4">
    <h3 class="p-2 text-left">
      <small><i class="fa-solid fa-sliders"></i> Adjust Parameters</small>
      <i
        id="collapse-parameters"
        class="fa-solid fa-circle-info info-clickable float-right"
      ></i>
    </h3>
    <div
      class="info-block mb-6"
      data-role="collapse"
      data-toggle-element="#collapse-parameters"
      data-collapsed="true"
    >
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent"
            >minimal variant call position coverage</span
          >
          The minimal number of reads to cover a variant position in order to be
          accepted. Requires a positive integer.
        </p>
      </div>
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent">minimal variant call quality</span>
          The minimal cell value of the QUAL field of provided
          <a
            href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531692-VCF-Variant-Call-Format"
            target="_blank"
            >VCF</a
          >
          files variants in order to be accepted. Requires a positive integer.
        </p>
      </div>
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent"
            >minimal homozygous variant call frequency</span
          >
          The minimal allele frequency (in %) of a called homozygous variant in
          order to be accepted. Requires a value between 0 and 100.
        </p>
      </div>
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent"
            >minimal heterozygous variant call frequency</span
          >
          The minimal allele frequency (in %) of a called heterozygous variant
          in order to be accepted. Requires a value between 0 and 100.
        </p>
      </div>
      <div class="p-5 text-left">
        <p>
          <span class="tag-translucent"
            >maximal heterozygous variant call frequency</span
          >
          The maximal allele frequency (in %) of a called heterozygous variant
          in order to be accepted. Requires a value between 0 and 100.
        </p>
      </div>
    </div>

    <table class="table">
      <tbody>
        <tr>
          <td class="w-75 text-left">Minimal Variant Call Position Coverage</td>
          <td>
            <input
              id="input-parameter-min-coverage"
              type="text"
              data-role="input"
              data-append="Position Coverage"
              data-default-value="5"
            />
          </td>
        </tr>
        <tr>
          <td class="w-75 text-left">
            Minimal Variant Call Quality (Phred Scaled Error Probability)
          </td>
          <td>
            <input
              id="input-parameter-min-quality"
              type="text"
              data-role="input"
              data-append="Phred Score"
              data-default-value="30"
            />
          </td>
        </tr>
        <tr>
          <td class="w-75 text-left">
            Minimal Homozygous Variant Call Frequency
          </td>
          <td>
            <input
              id="input-parameter-min-hom-frequency"
              type="text"
              data-role="input"
              data-append="%"
              data-default-value="90"
            />
          </td>
        </tr>
        <tr>
          <td class="w-75 text-left">
            Minimal Heterozygous Variant Call Frequency
          </td>
          <td>
            <input
              id="input-parameter-min-het-frequency"
              type="text"
              data-role="input"
              data-append="%"
              data-default-value="45"
            />
          </td>
        </tr>
        <tr>
          <td class="w-75 text-left">
            Maximal Heterozygous Variant Call Frequency
          </td>
          <td>
            <input
              id="input-parameter-max-het-frequency"
              type="text"
              data-role="input"
              data-append="%"
              data-default-value="55"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>

  <div class="remark m-4">
    <button
      onclick="submit()"
      id="upload-data-button"
      class="button small rounded text-upper"
      data-role="hint"
      data-hint-text="Submit your request to the server."
      data-hint-position="right"
      disabled
    >
      submit request
    </button>
    <span
      ><sup class="mandatory-tag">*</sup
      ><sup class="toolbar-text">Mandatory fields must be filled in.</sup></span
    >
  </div>
</div>
{% endblock %}